proposal/package-importer.d.ts.md

```diff
@@ -0,0 +1,507 @@
+JavaScript entry point. Some packages use the `"sass"` key at the root of their
+`package.json`.
+
+Other packages have adopted [conditional exports], driven by build tools like
+[Vite], [Parcel] and [Sass Loader for Webpack] which all resolve Sass paths
+using the `"sass"` and the `"style"` custom conditions.
+
+[conditional exports]: https://nodejs.org/api/packages.html#conditional-exports
+[Vite]: https://github.com/vitejs/vite/pull/7817
+[Parcel]: https://github.com/parcel-bundler/parcel/blob/2d2400ded4615375ee6bd53ef77b4857ad1591dd/packages/transformers/sass/src/SassTransformer.js#L163
+[Sass Loader for Webpack]: https://github.com/webpack-contrib/sass-loader/blob/02df41203adfda96959e56abb43bd35a89ec11ba/src/utils.js#L514
+
+Because use of conditional exports is flexible and recommended for modern
+packages, this will be the primary method used for the Node package importer. We
+will support both the `"sass"` and the `"style"` conditions, as Sass can also
+use the CSS exports exposed through `"style"`. While in practice, `"style"`
+tends to be used solely for `css` files, we will support `scss`, `sass` and
+`css` files for either `"sass"` or `"style"`.
+
+While conditional exports allows package authors to define specific aliases to internal
+files, we will still use the Sass conventions for resolving file paths with
+partials, extensions and indices to discover the intended export alias. However,
+we will not apply that logic to the destination, and will expect library authors
+to map the export to the correct place. In other words, given a `package.json`
+with `exports` as below, The Node package importer will resolve a
+`@use "pkg:pkgName/variables";` to the destination of the `_variables.scss` export.
+
+```json
+{
+  "exports": {
+    "_variables.scss": {
+      "sass": "./src/sass/_variables.scss"
+    }
+  }
+}
+```
+
+## Types
+
+```ts
+import '../spec/js-api';
+```
+
+### `pkgImporter`
+
+If set, the compiler will use the specified built-in package importer to resolve
+any URL with the `pkg:` scheme. Currently, the only available package importer
+is `node`, which follows Node resolution logic to locate Sass files.
+
+Defaults to undefined.
+
+After the first bullet points in [`compile`] and [`compileString`] in the
+Javascript Compile API, insert:
+
+* If `options.pkgImporter` equals `'node'`:
+
+  * Let `pkgImporter` be a [Node Package Importer] with an associated
+    `entryPointURL` of `require.main.filename`.
+
+  * Append `pkgImporter` to the `options.importers`.
+  
+  > Package Importers are evaluated after user-defined importers but
+  > before load paths.
+
+[`compile`]: ../spec/js-api/compile.d.ts.md#compile
+[`compileString`]: ../spec/js-api/compile.d.ts.md#compilestring
+[Node Package Importer]: #node-package-importer
+
+```ts
+declare module '../spec/js-api/options' {
+  interface Options<sync extends 'sync' | 'async'> {
+    pkgImporter?: 'node';
+  }
```

Comment:
My concern is that this design is a limited to a single pkg importer. E.g. In ruby implementation it is likely I want to ship two `pkgImporter`s, one for `node` and one for `gem`. Also, I'd like to be able to enable multiple pkg importers with user defined order.

If we are going to mutate `options.importers` anyways, we can ship a `NodePkgImporter` stub singleton class, which mocks the importer interface, but do not contain actual implementation. This way users can directly put `NodePkgImporter` into `options.importers` array in the desired position. When we see `NodePkgImporter` as we process the `options.importers`, we can internally swap the stub class to an actual implementation, or for embedded-sass we can swap that with the embedded protocol constant.

For example:

``` ruby
Sass.compile_string(
  '@import "pkg:bootstrap";',
   importers: [Sass::NodePkgImporter, Sass::GemPkgImporter, ...other_importers]
)
```

@nex3 What do you think?

Reply:
The main question I have here is, what's the use-case for having multiple package importers that coexist in the same compilation? Is it realistic that users would want to try loading from `node_modules` and then fall back to RubyGems?

Reply:
It is probably rare in the real world usage, but it is possible for a certain package to be only available in one package manager or another, that in some cases user will have to mix two pkg importers to load different packages.

Reply:
All right, I think that's reasonable.

@jamesnw Can you update this to change the modern API from using a flag to providing an opaque `nodePkgImporter` object? [This stackoverflow issue](https://stackoverflow.com/questions/56737033/how-to-define-an-opaque-type-in-typescript) provides a suggestion for how to define such a typing. The actual object itself can just be a `new Object()` that we compare with `===`.

Reply:
I've implemented that here- [6cdf76a](https://github.com/sass/sass/pull/3660/commits/6cdf76a3d2c268ef2c5be53790213af6bf5ed668)

It currently is not valid Typescript, as [Declaration Merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#merging-interfaces) doesn't work if non-function types are not unique or the same type. Here, we are needing to completely _replace_ the type of `Options.importers` to add the new NodePackageImporter type. 

The error it's showing is `Subsequent property declarations must have the same type.` I'll continue digging, but if there's an example of where we've replaced a type elsewhere, that would be helpful.

Reply:
@nex3 After some investigation, I think our 2 best options are-

1. Introduce a new option called something better than `importers2` that allows for Importer, FileImporter, or NodePackageImporter. 
2. Decide to overwrite rather than merge the new definition. This would be a non-breaking, additive change, but still require mutating the original spec directly.

Reply:
Ah, we've finally reached the limits on TypeScript's ability to do what we want. Let's call it `importers_new_` to distinguish it, I suppose.

Reply:
Updated in https://github.com/sass/sass/pull/3660/commits/827855a369e213d07d32978b4679dd199103537a

This should resolve the Static Analysis and TypeDoc build failures in the CI. The Link Referents failure is for a link to  https://runtime-keys.proposal.wintercg.org/#adding-a-key that is valid, but appears to be blocking requests from Github. Unless there's some workaround, it may be easiest to remove the link.

Reply:
I wonder for the JS API why cannot it just be an Importer type on the public API?

e.g.

``` js
export const nodePkgImporter = {
  canonicalize: function () {
    throw 'This is only a stub that is not meant to be invoked by user';
  },
  load: function () {
    throw 'This is only a stub that is not meant to be invoked by user';
  }
}
```

``` js
if (importer === sass.nodePkgImporter) {
  // enable the internal implementation
}
```

Reply:
I like this solution- it allows us to keep the simpler ‘importers’ option and doesn’t rely on the undocumented Typescript conventions for opaque objects. @nex3 would this work for you?

Reply:
In talking with @nex3,  we decided to use the opaque option to make it more clear that it's not something a user can create. I also got clarification that the `importers_new_` was only for declaration merging in the spec, but on implementation, that would replace the existing `importers` option, and not introduce a new option to the API. Implemented in https://github.com/sass/sass/pull/3660/commits/c8fcd99b1d870041116b931618d996c9aafea0b3